src/hotspot/cpu/aarch64/aarch64_neon_ad.m4

```diff
@@ -2048,18 +2048,21 @@ instruct vsra$3$4_imm`'(vec$6 dst, vec$6 src, immI shift) %{
   match(Set dst (RShiftV$4 src (RShiftCntV shift)));
   ins_cost(INSN_COST);
   format %{ "$1    $dst, $src, $shift\t# vector ($3$5)" %}
-  ins_encode %{ifelse($4, B,`
-    int sh = (int)$shift$$constant;
-    if (sh >= 8) sh = 7;
-    __ $2(as_FloatRegister($dst$$reg), __ T$3$5,
-           as_FloatRegister($src$$reg), sh);', $4, S,`
+  ins_encode %{
     int sh = (int)$shift$$constant;
-    if (sh >= 16) sh = 15;
-    __ $2(as_FloatRegister($dst$$reg), __ T$3$5,
-           as_FloatRegister($src$$reg), sh);', `
-    __ $2(as_FloatRegister($dst$$reg), __ T$3$5,
-            as_FloatRegister($src$$reg),
-            (int)$shift$$constant);')
+    if (sh == 0) {
+      __ orr(as_FloatRegister($dst$$reg), __ ifelse($6, D, T8B, T16B),
+             as_FloatRegister($src$$reg), as_FloatRegister($src$$reg));
+    } else {ifelse($4, B,`
+      if (sh >= 8) sh = 7;
```

Comment:
I think it would be possible to move some of this logic from the AD file into MacroAssembler, with macros to generate the appropriate instruction based on their arguments. This might be cleaner: the logic here is very hard to follow.

Reply:
I backed out the modifications of `aarch64_neon.ad` and `aarch64_neon_ad.m4`.
The `shift == 0` case is handled by the assembler now. Verified with the regression tests.

Reply:
> I think it would be possible to move some of this logic from the AD file into MacroAssembler, with macros to generate the appropriate instruction based on their arguments. This might be cleaner: the logic here is very hard to follow.

Hi, I moved the logic to the assembler.
The assembler will generate different instructions based on the value of `shift`.
If `shift == 0` and need not to accumulte, generated a `mov`.
If `shift == 0` and need to accumulte, generated an `add`.

Also added tests in `aarch64-asmtest.py` to verify the assembler modifications.